# 4.4: Data Wrangling & Subsetting 2

# 01. Importing Libraries

In [1]:
# Import libraries
import pandas as pd
import numpy as np
import os


# 02 Import Data

In [2]:
path = r'C:\Users\tee3a\3D Objects\12-2022 Instacart Basket Analysis'

In [3]:
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Original Data for raw data sets', 'orders.csv'), index_col = False)

In [4]:
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data for raw data sets', 'products.csv'), index_col = False)

In [5]:
# Dropping columns
df_ords.drop(columns = ['eval_set'])


order_id  user_id  order_number  order_dow  order_hour_of_day  \
0         2539329        1             1          2                  8   
1         2398795        1             2          3                  7   
2          473747        1             3          3                 12   
3         2254736        1             4          4                  7   
4          431534        1             5          4                 15   
...           ...      ...           ...        ...                ...   
3421078   2266710   206209            10          5                 18   
3421079   1854736   206209            11          4                 10   
3421080    626363   206209            12          1                 12   
3421081   2977660   206209            13          1                 12   
3421082    272231   206209            14          6                 14   

         days_since_prior_order  
0                           NaN  
1                          15.0  
2                          21.0  
3                          29.0  
4                          28.0  
...                         ...  
3421078                    29.0  
3421079                    30.0  
3421080                    18.0  
3421081                     7.0  
3421082                    30.0  

[3421083 rows x 6 columns]

In [6]:
df_ords_2 = df_ords.drop(columns = ['eval_set'])

In [7]:
df_ords['days_since_prior_order'].value_counts(dropna = False)

30.0    369323
7.0     320608
6.0     240013
4.0     221696
3.0     217005
5.0     214503
NaN     206209
2.0     193206
8.0     181717
1.0     145247
9.0     118188
14.0    100230
10.0     95186
13.0     83214
11.0     80970
12.0     76146
0.0      67755
15.0     66579
16.0     46941
21.0     45470
17.0     39245
20.0     38527
18.0     35881
19.0     34384
22.0     32012
28.0     26777
23.0     23885
27.0     22013
24.0     20712
25.0     19234
29.0     19191
26.0     19016
Name: days_since_prior_order, dtype: int64

In [8]:
# Renaming Columns
df_ords_2.rename(columns = {'order_dow' : 'orders_day_of_week'}, inplace = True)

In [9]:
df_ords_2.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [10]:
df_ords_2['order_id'] = df_ords_2['order_id'].astype('str')

In [11]:
df_ords_2['user_id'] = df_ords_2['user_id'].astype('str')

In [12]:
df_ords_2.describe()

order_number  orders_day_of_week  order_hour_of_day  \
count  3.421083e+06        3.421083e+06       3.421083e+06   
mean   1.715486e+01        2.776219e+00       1.345202e+01   
std    1.773316e+01        2.046829e+00       4.226088e+00   
min    1.000000e+00        0.000000e+00       0.000000e+00   
25%    5.000000e+00        1.000000e+00       1.000000e+01   
50%    1.100000e+01        3.000000e+00       1.300000e+01   
75%    2.300000e+01        5.000000e+00       1.600000e+01   
max    1.000000e+02        6.000000e+00       2.300000e+01   

       days_since_prior_order  
count            3.214874e+06  
mean             1.111484e+01  
std              9.206737e+00  
min              0.000000e+00  
25%              4.000000e+00  
50%              7.000000e+00  
75%              1.500000e+01  
max              3.000000e+01

## Transposing Data

In [13]:
# Transposing Data
df_dep = pd.read_csv(os.path.join(path, '02 Data', 'Original Data for raw data sets', 'products.csv'), index_col = False)

In [14]:
df_dep.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

## Task 4.4

In [15]:
#  2. Find another identifier variable in the df_ords dataframe that doesn’t need to be included in your analysis as a numeric variable and change it to a suitable format.
df_ords_2['order_number'] = df_ords_2['order_number'].astype('str')

In [16]:
df_ords_2.describe()

orders_day_of_week  order_hour_of_day  days_since_prior_order
count        3.421083e+06       3.421083e+06            3.214874e+06
mean         2.776219e+00       1.345202e+01            1.111484e+01
std          2.046829e+00       4.226088e+00            9.206737e+00
min          0.000000e+00       0.000000e+00            0.000000e+00
25%          1.000000e+00       1.000000e+01            4.000000e+00
50%          3.000000e+00       1.300000e+01            7.000000e+00
75%          5.000000e+00       1.600000e+01            1.500000e+01
max          6.000000e+00       2.300000e+01            3.000000e+01

In [17]:
# 3. Look for a variable in your df_ords dataframe with an unintuitive name and change its name without overwriting the data frame.

df_ords_2.rename(columns = {'days_since_prior_order' : 'days_between_order'}, inplace = True)

In [18]:
df_ords_2.head()

order_id user_id order_number  orders_day_of_week  order_hour_of_day  \
0  2539329       1            1                   2                  8   
1  2398795       1            2                   3                  7   
2   473747       1            3                   3                 12   
3  2254736       1            4                   4                  7   
4   431534       1            5                   4                 15   

   days_between_order  
0                 NaN  
1                15.0  
2                21.0  
3                29.0  
4                28.0

In [19]:
#4. Your client wants to know what the busiest hour is for placing orders. Find the frequency of the corresponding variable and share your findings.
df_ords_2.describe()

orders_day_of_week  order_hour_of_day  days_between_order
count        3.421083e+06       3.421083e+06        3.214874e+06
mean         2.776219e+00       1.345202e+01        1.111484e+01
std          2.046829e+00       4.226088e+00        9.206737e+00
min          0.000000e+00       0.000000e+00        0.000000e+00
25%          1.000000e+00       1.000000e+01        4.000000e+00
50%          3.000000e+00       1.300000e+01        7.000000e+00
75%          5.000000e+00       1.600000e+01        1.500000e+01
max          6.000000e+00       2.300000e+01        3.000000e+01

In [20]:
df_ords_2['order_hour_of_day'].value_counts(dropna = False)

10    288418
11    284728
15    283639
14    283042
13    277999
12    272841
16    272553
9     257812
17    228795
18    182912
8     178201
19    140569
20    104292
7      91868
21     78109
22     61468
23     40043
6      30529
0      22758
1      12398
5       9569
2       7539
4       5527
3       5474
Name: order_hour_of_day, dtype: int64

10 is the busiest hour is for placing orders total of 288, 418

In [21]:
#5 Determine the meaning behind a value of 4 in the "department_id" column within the df_prods dataframe using a data dictionary.
data_dict_prods = df_prods.to_dict('index')

In [22]:
df_produce =  df_prods[df_prods['department_id']==4]

In [23]:
df_produce.head()

product_id                    product_name  aisle_id  department_id  \
30          31              White Pearl Onions       123              4   
42          43             Organic Clementines       123              4   
44          45               European Cucumber        83              4   
65          66       European Style Spring Mix       123              4   
88          89  Yogurt Fruit Dip Sliced Apples       123              4   

    prices  
30     7.5  
42    11.5  
44    14.3  
65    11.7  
88    12.6

In [24]:
# 6. The sales team in your client’s organization wants to know more about breakfast item sales. Create a subset containing only the required information.

df_breakfast = df_prods.loc[df_prods['department_id'] == 14]

In [25]:
df_breakfast.head()

product_id                                      product_name  aisle_id  \
27           28                                 Wheat Chex Cereal       121   
33           34                                               NaN       121   
67           68                           Pancake Mix, Buttermilk       130   
89           90                                      Smorz Cereal       121   
210         211  Gluten Free Organic Cereal Coconut Maple Vanilla       130   

     department_id  prices  
27              14    10.1  
33              14    12.2  
67              14    13.7  
89              14     3.9  
210             14     3.6

In [26]:
# 7. Observations from the entire dataframe that include items from the following departments: alcohol, deli, beverages, and meat/seafood. 
df_alcohol =  df_prods[df_prods['department_id']==5]

In [27]:
df_deli =  df_prods[df_prods['department_id']==20]

In [28]:
df_beverages =  df_prods[df_prods['department_id']==7]

In [29]:
df_meatseafood =  df_prods[df_prods['department_id']==12]

In [30]:
df_inquiries = df_alcohol + df_deli + df_beverages + df_meatseafood

In [31]:
df_inquiries

product_id product_name  aisle_id  department_id  prices
2             NaN          NaN       NaN            NaN     NaN
6             NaN          NaN       NaN            NaN     NaN
9             NaN          NaN       NaN            NaN     NaN
10            NaN          NaN       NaN            NaN     NaN
16            NaN          NaN       NaN            NaN     NaN
...           ...          ...       ...            ...     ...
49676         NaN          NaN       NaN            NaN     NaN
49679         NaN          NaN       NaN            NaN     NaN
49680         NaN          NaN       NaN            NaN     NaN
49686         NaN          NaN       NaN            NaN     NaN
49688         NaN          NaN       NaN            NaN     NaN

[7650 rows x 5 columns]

The observation noticed here is that there is no customer buying all these products; alcohol, deli, beverages, and meat/seafood, at the same time. So, there is no customers who might be throwing dinner parties.

# 8. How many rows does the last dataframe you created have?
7650 rows

In [32]:
data_dict_ords = df_ords_2.to_dict('index')

In [33]:
data_dict_ords

{0: {'order_id': '2539329',
  'user_id': '1',
  'order_number': '1',
  'orders_day_of_week': 2,
  'order_hour_of_day': 8,
  'days_between_order': nan},
 1: {'order_id': '2398795',
  'user_id': '1',
  'order_number': '2',
  'orders_day_of_week': 3,
  'order_hour_of_day': 7,
  'days_between_order': 15.0},
 2: {'order_id': '473747',
  'user_id': '1',
  'order_number': '3',
  'orders_day_of_week': 3,
  'order_hour_of_day': 12,
  'days_between_order': 21.0},
 3: {'order_id': '2254736',
  'user_id': '1',
  'order_number': '4',
  'orders_day_of_week': 4,
  'order_hour_of_day': 7,
  'days_between_order': 29.0},
 4: {'order_id': '431534',
  'user_id': '1',
  'order_number': '5',
  'orders_day_of_week': 4,
  'order_hour_of_day': 15,
  'days_between_order': 28.0},
 5: {'order_id': '3367565',
  'user_id': '1',
  'order_number': '6',
  'orders_day_of_week': 2,
  'order_hour_of_day': 7,
  'days_between_order': 19.0},
 6: {'order_id': '550135',
  'user_id': '1',
  'order_number': '7',
  'orders_day_o

In [34]:
df_ords_2['user_id'].value_counts(dropna = False)

152340    100
185641    100
185524    100
81678     100
70922     100
         ... 
13046       4
70208       4
111440      4
31762       4
47980       4
Name: user_id, Length: 206209, dtype: int64

In [35]:
df_orders_user_id_1_information = df_ords_2.loc[df_ords['user_id'] == 1]      

In [36]:
# 9. All the information you can about this user_id 1.

df_orders_user_id_1_information

order_id user_id order_number  orders_day_of_week  order_hour_of_day  \
0   2539329       1            1                   2                  8   
1   2398795       1            2                   3                  7   
2    473747       1            3                   3                 12   
3   2254736       1            4                   4                  7   
4    431534       1            5                   4                 15   
5   3367565       1            6                   2                  7   
6    550135       1            7                   1                  9   
7   3108588       1            8                   1                 14   
8   2295261       1            9                   1                 16   
9   2550362       1           10                   4                  8   
10  1187899       1           11                   4                  8   

    days_between_order  
0                  NaN  
1                 15.0  
2                 21.0  
3                 29.0  
4                 28.0  
5                 19.0  
6                 20.0  
7                 14.0  
8                  0.0  
9                 30.0  
10                14.0

In [37]:
df_orders_user_id_1_information

order_id user_id order_number  orders_day_of_week  order_hour_of_day  \
0   2539329       1            1                   2                  8   
1   2398795       1            2                   3                  7   
2    473747       1            3                   3                 12   
3   2254736       1            4                   4                  7   
4    431534       1            5                   4                 15   
5   3367565       1            6                   2                  7   
6    550135       1            7                   1                  9   
7   3108588       1            8                   1                 14   
8   2295261       1            9                   1                 16   
9   2550362       1           10                   4                  8   
10  1187899       1           11                   4                  8   

    days_between_order  
0                  NaN  
1                 15.0  
2                 21.0  
3                 29.0  
4                 28.0  
5                 19.0  
6                 20.0  
7                 14.0  
8                  0.0  
9                 30.0  
10                14.0

# 10. What basic stats can you provide based on the information you have about user_id 1?
User id 1 made a total of 11 orders, with different order ids and with the order numbers ranging from 1 to 11.User id 1 also ordered from day 1 to day 4 of the week. The order hours are 7,8,9,12, 14, 15 and 16. Lastly, order days between is missing an information, while the remaining 10 information range between 0 and 29. 

# Exporting data

In [38]:
df_ords_2.to_csv(os.path.join(path, '02 Data','Prepared Data for storing any manipulated data sets', 'orders_wrangled.csv'))

In [39]:
df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data for storing any manipulated data sets', 'orders_wrangled.csv'))

In [40]:
df_ords_2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3421083 entries, 0 to 3421082
Data columns (total 6 columns):
 #   Column              Dtype  
---  ------              -----  
 0   order_id            object 
 1   user_id             object 
 2   order_number        object 
 3   orders_day_of_week  int64  
 4   order_hour_of_day   int64  
 5   days_between_order  float64
dtypes: float64(1), int64(2), object(3)
memory usage: 156.6+ MB
